### PR TITLE
WIP boost scores based on atoms

### PR DESCRIPTION
**Pull request by keegancsmith**
Early preview of this work. I am playing around with factors/etc. Also only works for ChunkMatches at the moment.

Some things it makes worse. For the atom boost queries we added, it improves all of those.

### CODE REVIEW

**Comment by jtibshirani** [comment]
Could you give some background on the overall approach? When we last chatted about ranking in our offsite, we talked about summing all the atom scores (and normalizing by the number of atoms).optionally normalizing them)

**Comment by jtibshirani** [comment]
Could you give some background on the overall approach? When we last chatted about ranking in our offsite, we talked about summing all the atom scores (and then normalizing by the number of atoms).

File: internal/e2e/testdata/bufio_flush_writer.txt [hunk 1]
```diff
@@ -1,41 +1,41 @@
 queryString: bufio flush writer
 query: (and substr:"bufio" substr:"flush" substr:"writer")
-targetRank: 25
+targetRank: 1
```
Comment (stefanhengl) [inline internal/e2e/testdata/bufio_flush_writer.txt]: nice!

**Comment by keegancsmith** [comment]
@stefanhengl @jtibshirani tl;dr not happy with this yet, gonna try the score per atom approach now. But after today I wanna take a pause on this work before resuming it before branch cut.

Alright I played around a bunch with this. The summary of what this does is it adds in a new scoring component which is based purely on the atoms your candidateMatches came from.

Score of a []*candidateMatch added in is:
- the % of content atoms from the matchTree whcih contributes to that slice.
- the max run of "increasing" atoms. This is so we can reward for a query "foo bar" that content which is "foo blah bar" higher than "bar blah foo".
- always mix in the % of filename atoms that match.

I experimented a bunch with different weights, but the problem I am running into is its hard to keep rewarding high quality symbol matches with trying to boost non-symbol matches that match the query nicely. The current state of the PR is where I am drawing a line in the sand.

I am not happy with this though, and maybe should decrease the contribution of queryAtoms further. Main reason I am not happy is important queries like "test server" no longer rank the http test server highly. Maybe take a look and see what you think.

In the meantime I am gonna prototype calculating a score per atom, and then combining. A lot of the work I did here should hopefully make that easier. But after today I am gonna take a pause on this work to tackle other stuff so it can stew a bit before I merge something final in.

**Comment by keegancsmith** [comment]
Phrase boosting solved our issues. May revisit these ideas in the future.

**Pull request closed without merging**